Cleanup of sc-ossl-compat.h for OpenSSL versions 1.1.1 and above

dengert opened the pull request:
With discussion to only support OpenSSL 1.1.1 and 3.0 above and
most of sc-ossl-compat.h can be removed.

This also removes support for LibreSSL.

 On branch improved-sc-ossl-compat
 Changes to be committed:
	modified:   sc-ossl-compat.h
	modified:   ../pkcs11/framework-pkcs15.c

<!--
Thank you for your pull request.

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX'
(without quotes) in the commit message.

Mention which card(s) are used during testing. To get the name of your card,
run this command: `opensc-tool -n`
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [ ] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

Jakuje commented:
This looks like it breaks osx build. Can you check what openssl version is used there?
```
ctx.c:848:7: error: implicit declaration of function 'CRYPTO_secure_malloc_initialized' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
        if (!CRYPTO_secure_malloc_initialized()) {
             ^
ctx.c:849:3: error: implicit declaration of function 'CRYPTO_secure_malloc_init' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                CRYPTO_secure_malloc_init(OPENSSL_SECURE_MALLOC_SIZE, OPENSSL_SECURE_MALLOC_SIZE/8);
                ^
ctx.c:849:3: note: did you mean 'CRYPTO_secure_malloc_initialized'?
ctx.c:848:7: note: 'CRYPTO_secure_malloc_initialized' declared here
        if (!CRYPTO_secure_malloc_initialized()) {
             ^
2 errors generated.
```
Similarly the windows builds in appveyor:
```
ctx.c
ctx.c(848): error C2220: warning treated as error - no 'object' file generated
ctx.c(848): warning C4013: 'CRYPTO_secure_malloc_initialized' undefined; assuming extern returning int
ctx.c(849): warning C4013: 'CRYPTO_secure_malloc_init' undefined; assuming extern returning int
```

Jakuje commented:
In any case, I would wait with this one after #2495 so we can see in which state are we in the LibreSSL waters. None of us has strong need to support LibreSSL, but it looks like it will not be that complicated in the end so if Frank can bring the PR to working stage with CI, we should be able to keep up with LibreSSL.

dengert commented:
The `CRYPTO_secure_malloc_initialized`, `CRYPTO_secure_malloc_initialized` and `CRYPTO_secure_malloc_initialized` were all deleted by this PR. They could be added back in. The were add as inline functions, but it looks like they could have been added 
as defines or macros:  `CRYPTO_secure_malloc_initialized` and  `CRYPTO_secure_malloc_initialized` and return 0. `CRYPTO_secure_malloc_initialized` is void. 

They where defined if this is true: `#if OPENSSL_VERSION_NUMBER < 0x10100000L || (defined(LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER < 0x2050300fL)` 

The are not listed in LibreSSL includes from 3.4.2, which has: `#define LIBRESSL_VERSION_NUMBER 0x3040100fL`   OpenSSL > 1.0.1 defines them and LibreSSL does not. So above `#if` is not correct.  

We will see how #2495 goes which says:
> it's a low hanging fruit and there actually are some distros which use LibreSSL.

The tree is growing fast...

mouse07410 commented:
I don't think LibreSSL is a "low-hanging" fruit, and don't think it's worth time and effort.

Jakuje commented:
Can you rebase the branch on top of the current master and fix the build issues?

dengert commented:
Yes but not till next week.

On Fri, Jan 28, 2022, 10:16 AM Jakub Jelen ***@***.***> wrote:

> Can you rebase the branch on top of the current master and fix the build
> issues?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenSC/OpenSC/pull/2492#issuecomment-1024317699>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGTIMONNMQPG2GVZHMSXILUYKXNHANCNFSM5MK3H2JQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

dengert commented:
The PR as submitted, dropped all support for LibreSSL and OpenSSL below 1.1.1 which allowed most of code to be deleted. (removed 242 lines and added 7).  

But https://github.com/OpenSC/OpenSC/pull/2495 has updated support for LibreSSL so I am closing 2492 and may submit a new PR to clean up parts of `sc-ossl-compat.h` in the future.
